lib/elixir/lib/module/checker.ex

```diff
@@ -0,0 +1,184 @@
+defmodule Module.Checker do
+  def verify(module_map, _binary) do
+    state = %{
+      file: module_map.file,
+      module: module_map.module,
+      compile_opts: module_map.compile_opts,
+      function: nil
+    }
+
+    module_map.definitions
+    |> check_definitions(state)
+    |> List.flatten()
+    |> merge_warnings()
+    |> sort_warnings()
+    |> emit_warnings()
+  end
+
+  defp check_definitions(definitions, state) do
+    Enum.map(definitions, &check_definition(&1, state))
+  end
+
+  defp check_definition({function, _kind, meta, clauses}, state) do
+    state = put_file_meta(%{state | function: function}, meta)
+    Enum.map(clauses, &check_clause(&1, state))
+  end
+
+  defp put_file_meta(state, meta) do
+    case Keyword.fetch(meta, :file) do
+      {:ok, {file, _}} -> %{state | file: file}
+      :error -> state
+    end
+  end
+
+  defp check_clause({_meta, _args, _guards, body}, state) do
+    check_body(body, state)
+  end
+
+  defp check_body({:&, meta, [{:/, _, [{{:., dot_meta, [module, fun]}, _, []}, arity]}]}, state)
+       when is_atom(module) and is_atom(fun) do
+    check_remote(module, fun, arity, meta ++ dot_meta, state)
+  end
+
+  defp check_body({{:., meta, [module, fun]}, _, args}, state)
+       when is_atom(module) and is_atom(fun) do
+    check_remote(module, fun, length(args), meta, state)
+  end
+
+  defp check_body({left, _meta, right}, state) when is_list(right) do
```

Comment:
Can you remind me what this AST is for? Maybe a comment would be nice. Actually we usually put a comment above these clauses on AST formats to show the original code that would end up there, which I think is nice :)

Reply:
Can you link to an example of these comments?

Reply:
Yep, a quick one: https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/code/formatter.ex#L497-L505 (and subsequent lines)

Similar case: https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/macro.ex#L692-L724

Reply:
I am not sure it adds much in this case because they are pretty general or self-explanatory except for maybe the two first ones.

Reply:
I added a commit with comments.

Reply:
That's perfect 💟